lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -0,0 +1,159 @@
+  if (configs.size() > 0) {
+    backend_.reset(createBackend(configs[0].backendKind));
+  }
+  for (auto &config : configs) {
+    auto newDevice = DeviceManager::createDeviceManager(config.backendKind,
+                                                        config.deviceName);
+    devices_.emplace(deviceCount, newDevice);
+    deviceCount++;
+  }
+
+  executor_.reset(createExecutor(devices_));
+}
+
+HostManager::~HostManager() { clearHost(); }
+
+ResultCode HostManager::addNetwork(Module *M) {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  std::vector<DeviceInfo> deviceInfo;
+  for (auto &device : devices_) {
+    DeviceInfo info = DeviceInfo();
+    info.availableMemory = device.second->getAvailableMemory();
+    deviceInfo.push_back(info);
+  }
+  // Optimize functions before passing to partitioner.
+  // Currently hardcoding inference.
+  if (backend_) {
+    for (auto F : M->getFunctions()) {
+      backend_->optimizeFunction(CompilationMode::Infer, F);
+    }
+  }
+  auto partitioner = Partitioner(M, deviceInfo);
+  // auto provisioner = llvm::make_unique<Provisioner>();
+  auto nodeList = std::move(partitioner.Partition());
+  auto result = provisioner_->provision(nodeList.roots, devices_, *M);
+
+  for (auto &node : nodeList.roots) {
+    roots_.emplace(node->name, std::move(node));
+  }
+  for (auto &node : nodeList.nodes) {
+    networks_.emplace(node->name, std::move(node));
+  }
+  return result;
+}
+
+void HostManager::removeNetwork(llvm::StringRef networkName) {
+  // Walk the tree for the given function, calling evict function for each node
+  // before removing it from networks_ which frees the node.
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  if (roots_.find(networkName) == roots_.end()) {
+    return;
+  }
+  std::queue<std::string> nodes;
+  std::set<std::string> allNodes;
+  for (auto &child : roots_[networkName]->children) {
+    nodes.push(child->name);
+    allNodes.insert(child->name);
+  }
+
+  while (!nodes.empty()) {
+    auto nodeName = nodes.front();
+    nodes.pop();
+    auto it = networks_.find(nodeName);
+    if (it != networks_.end()) {
+      for (auto &node : it->second->children) {
+        auto name = node->name;
+        if (allNodes.find(node->name) == allNodes.end()) {
+          nodes.push(node->name);
+          allNodes.insert(node->name);
+        }
+      }
+    }
+  }
+  roots_.erase(roots_.find(networkName));
+  for (auto &networkName : allNodes) {
+    auto networkIterator = networks_.find(networkName);
+    if (networkIterator != networks_.end()) {
+      devices_[networkIterator->second->deviceID]->evictNetwork(networkName);
+      networks_.erase(networkIterator);
+    }
+  }
+}
+
+bool HostManager::networkAdded(llvm::StringRef networkName) {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  return roots_.find(networkName) != roots_.end();
+}
+
+void HostManager::clearHost() {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
```

Comment:
does lock need to be held while stopping devices?

Reply:
For the device no, but this way we can't add a function to a device we are stopping.